quote-book/synchronized/src/lib.rs

```diff
@@ -82,30 +145,245 @@ where
+            highest_processed_block_index,
+            stop_requested,
+            logger,
+        };
+        thread.run();
+    }
+
+    fn run(mut self) {
+        log::info!(self.logger, "Db fetcher thread started.");
+        loop {
+            if self.stop_requested.load(Ordering::SeqCst) {
+                log::info!(self.logger, "Db fetcher thread stop requested.");
+                break;
+            }
+            let starting_block_index = self.next_block_index;
+            while self.load_block_data() && !self.stop_requested.load(Ordering::SeqCst) {}
+            if self.next_block_index != starting_block_index {
+                let last_processed_block_index = self.next_block_index - 1;
+                let mut quotes = self
+                    .quotebook
+                    .remove_quotes_by_tombstone_block(last_processed_block_index);
+                while quotes.is_err() {
+                    log::error!(
+                        self.logger,
+                        "Unexpected error when removing quotes by tombstone_block {}. Retrying. Error: {}",
+                        last_processed_block_index,
+                        quotes.expect_err("quotes should be an err if we're printing about the unexpected error")
+                    );
+                    std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+                    quotes = self
+                        .quotebook
+                        .remove_quotes_by_tombstone_block(last_processed_block_index);
+                }
+                (self.remove_quote_callback.lock().expect("lock poisoned"))(quotes.unwrap());
+                self.highest_processed_block_index
+                    .store(last_processed_block_index, Ordering::SeqCst);
+            }
+            std::thread::sleep(Self::POLLING_FREQUENCY);
+        }
+    }
+
+    /// Attempt to load the next block that we
+    /// are aware of and remove quotes that match key images from it.
+    /// Returns true if we might have more block data to load.
+    fn load_block_data(&mut self) -> bool {
+        // Default to true: if there is an error, we may have more work, we don't know
+        let mut may_have_more_work = true;
+
+        match self.db.get_block_contents(self.next_block_index) {
+            Err(LedgerError::NotFound) => may_have_more_work = false,
+            Err(e) => {
+                log::error!(
+                    self.logger,
+                    "Unexpected error when checking for block data {}: {:?}",
+                    self.next_block_index,
+                    e
+                );
+                std::thread::sleep(Self::ERROR_RETRY_FREQUENCY);
+            }
+            Ok(block_contents) => {
+                // Filter keyimages in quotebook
+                for key_image in block_contents.key_images {
+                    let mut quotes = self.quotebook.remove_quotes_by_key_image(&key_image);
+                    while quotes.is_err() {
```

Comment:
I think this is broken, because the API treats "quote not found" as an error:

https://github.com/mobilecoinofficial/deqs/blob/90c98ebea832ac0b56dabca25605d450d17665cc/quote-book/api/src/error.rs#L21

we need to handle that explicitly and not loop when that happens

Reply:
if that's not the actual behavior of

```
https://github.com/mobilecoinofficial/deqs/blob/90c98ebea832ac0b56dabca25605d450d17665cc/quote-book/api/src/quote_book.rs#L25
```

then maybe we should just document how this trait function is supposed to behave if no matching quote is found

Reply:
I think because it returns the list of quotes that are removed every time, it shouldn't return quotenotfound errors, since it should instead just return an empty list of quotes to indicate this. I'll add to the description for this.